src/string.cr

```diff
@@ -3716,7 +3718,7 @@ class String
 
   # Searches the string for instances of *pattern*,
   # yielding a `Regex::MatchData` for each match.
-  def scan(pattern : Regex)
+  def scan(pattern : Regex, &block) : Nil
```

Comment:
`&block : Regex::MatchData ->`?

Reply:
adding this annotation makes the spec fail with:

```
Failures:

  1) scans using $~

       Nil assertion failed (NilAssertionError)
         from src/nil.cr:106:5 in 'not_nil!'
         from spec/std/string_spec.cr:1487:39 in '->'
         from src/spec/example.cr:255:3 in 'run'
         from src/spec/context.cr:207:23 in 'run'
         from src/spec/context.cr:42:23 in 'run'
         from src/spec/dsl.cr:162:7 in '->'
         from src/kernel.cr:255:3 in 'run'
         from src/crystal/main.cr:47:5 in 'main'
         from src/crystal/main.cr:106:3 in 'main'
         from __libc_start_main
         from _start
         from ???
```
‽

Reply:
@asterite any reason why adding this type restriction would break `$1`??

Reply:
A compiler bug